test/ad.jl

```diff
@@ -15,6 +15,10 @@ end
 function ChainRulesTestUtils.rand_tangent(rng::AbstractRNG, x::AbstractTensorMap)
     return randn!(similar(x))
 end
+function ChainRulesTestUtils.rand_tangent(rng::AbstractRNG, x::DiagonalTensorMap)
+    S = x.domain
+    return DiagonalTensorMap(randn(eltype(x), sum(values(S.dims))), S)
```

Comment:
I would prefer to not use `S` for a space but rather use `V`. Also, there is a function `reduceddim(V)` for `sum(values(V.dims))` (as the latter uses private implementation details).